prefect_bitbucket/repository.py

```diff
@@ -0,0 +1,186 @@
+"""
+Allows for interaction with a BitBucket repository.
+
+The `BitBucket` class in this collection is a storage block that lets Prefect agents
+pull Prefect flow code from BitBucket repositories.
+
+The `BitBucket` block is ideally configured via the Prefect UI, but can also be used
+in Python as the following examples demonstrate.
+
+Examples:
+```python
+    from prefect_bitbucket.repository import BitBucketRepository
+
+    # public BitBucket repository
+    public_bitbucket_block = BitBucketRepository(
+        repository="https://bitbucket.com/my-project/my-repository.git"
+    )
+
+    public_bitbucket_block.save(name="my-bitbucket-block")
+
+
+    # specific branch or tag
+    branch_bitbucket_block = BitBucketRepository(
+        reference="branch-or-tag-name",
+        repository="https://bitbucket.com/my-project/my-repository.git"
+    )
+
+    branch_bitbucket_block.save(name="my-bitbucket-block")
+
+
+    # private BitBucket repository
+    private_bitbucket_block = BitBucketRepository(
+        repository="https://bitbucket.com/my-project/my-repository.git",
+        access_token="MY_BITBUCKET_PERSONAL_ACCESS_TOKEN"
+    )
+
+    private_bitbucket_block.save(name="my-private-bitbucket-block")
+```
+
+"""
+
+import io
+from distutils.dir_util import copy_tree
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Optional, Tuple, Union
+from urllib.parse import urlparse, urlunparse
+
+from prefect.exceptions import InvalidRepositoryURLError
+from prefect.filesystems import ReadableDeploymentStorage
+from prefect.utilities.asyncutils import sync_compatible
+from prefect.utilities.processutils import run_process
+from pydantic import Field, validator
+
+from prefect_bitbucket.credentials import BitBucketCredentials
+
+
+class BitBucketRepository(ReadableDeploymentStorage):
+    """
+    Interact with files stored in BitBucket repositories.
+
+    An accessible installation of git is required for this block to function
+    properly.
+    """
+
+    _block_type_name = "BitBucket Repository"
+    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/27LMR24ewTSDW238Lks1vH/34c5028659f4007528feadc8db8cecbd/500px-Bitbucket-blue-logomark-only.svg.png?h=250"  # noqa
+    _description = "Interact with files stored in BitBucket repositories."
+
+    repository: str = Field(
+        default=...,
+        description="The URL of a BitBucket repository to read from in HTTPS format",
+    )
+    reference: Optional[str] = Field(
+        default=None,
+        description="An optional reference to pin to; can be a branch or tag.",
+    )
+    credentials: Optional[BitBucketCredentials] = Field(
```

Comment:
I am slightly leaning towards renaming this to `bitbucket_credentials` even though it's verbose since all our other Collections prefixes `credentials` with the collection name and it's clearer to the user that only "BitBucketCredentials" is allowed.

Reply:
Done